src/coreclr/jit/rationalize.cpp

```diff
@@ -265,6 +265,58 @@ void Rationalizer::RewriteIntrinsicAsUserCall(GenTree** use, ArrayStack<GenTree*
                       arg1, arg2);
 }
 
+#ifdef TARGET_ARM64
+// RewriteSubLshDiv: Possibly rewrite a SubLshDiv node into a Mod.
+//
+// Arguments:
+//    use - A use of a node.
+//
+// Transform: a - (a / cns) >> shift  =>  a % cns1
+//            where cns is an signed integer constant that is a power of 2.
+// We do this transformation because Lowering has a specific optimization
+// for 'a % cns1' that is not easily reduced by other means.
+//
+void Rationalizer::RewriteSubLshDiv(GenTree** use)
+{
+    GenTree* const node = *use;
+
+    if (!node->OperIs(GT_SUB))
+        return;
+
+    GenTree* op1 = node->gtGetOp1();
+    GenTree* op2 = node->gtGetOp2();
+
+    if (!(node->TypeIs(TYP_INT, TYP_LONG) && !node->IsUnsigned() && op1->OperIs(GT_LCL_VAR)))
+        return;
+
+    if (!op2->OperIs(GT_LSH))
+        return;
+
+    GenTree* lsh   = op2;
+    GenTree* div   = lsh->gtGetOp1();
+    GenTree* shift = lsh->gtGetOp2();
+    if (div->OperIs(GT_DIV) && shift->IsIntegralConst())
+    {
+        GenTree* a   = div->gtGetOp1();
+        GenTree* cns = div->gtGetOp2();
+        if (a->OperIs(GT_LCL_VAR) && cns->IsIntegralConstPow2() &&
+            op1->AsLclVar()->GetLclNum() == a->AsLclVar()->GetLclNum())
+        {
+            size_t shiftValue = shift->AsIntConCommon()->IntegralValue();
+            size_t cnsValue   = cns->AsIntConCommon()->IntegralValue();
+            if ((cnsValue >> shiftValue) == 1)
+            {
+                node->ChangeOper(GT_MOD);
+                node->AsOp()->gtOp2 = cns;
+
+                BlockRange().Remove(op2, shift);
```

Comment:
I don't think this is right if `lsh` has `GTF_REVERSE_OPS` set. I think to handle `GTF_REVERSE_OPS` properly the easiest will be to remove all the nodes individually.

It would be very helpful to add a comment showing the full tree that is being changed here.

Reply:
I'm basically going to bail if I see any of the ops having the reverse op set.

For adding a comment, is `a - (a / cns) << shift  =>  a % cns` not sufficient? I could add the full tree here, but it feels redundant.

Reply:
> For adding a comment, is a - (a / cns) << shift => a % cns not sufficient? I could add the full tree here, but it feels redundant.

I'll leave it up to you, I personally would prefer the detailed tree form but if you prefer the simplified form then that's totally fine with me.

I still do think the code here is unnecessarily complicated. Why not just remove only the nodes that you don't reuse? That way there is no need to resequence the tree or find the first node, plus no need to be conservative about `GTF_REVERSE_OPS` because removing the nodes individually handles that too. I.e. the entirety of the block becomes:
```cpp
node->ChangeOper(GT_MOD);
node->AsOp()->gtOp2 = cns;
BlockRange().Remove(lsh);
BlockRange().Remove(div);
BlockRange().Remove(a);
BlockRange().Remove(shift);
```
if I'm not mistaken.

BTW, should we guard this transform under `comp->opts.OptimizationEnabled()`?

Reply:
> I still do think the code here is unnecessarily complicated.

My head still has a hard time when it comes to lowering when I have to worry about the trees and the linear form at the same time; that in of itself is already complicated. So, I'm not surprised if what I have is complicated :) - you notice I keep going back and forth between different versions of trying to do the lowering. This is the reason why I was experimenting with a new phase right before rationalization where I would only have to worry about the trees; less to worry/think about.

> comp->opts.OptimizationEnabled()

Wouldn't hurt to do so.